add Heavy Light Decomposition

OmAmar106 opened the pull request:
Added Implementation of HLD using Segment Tree, since it was not present in the graphs section.

The Implementation gets AC on the problem : https://acm.timus.ru/problem.aspx?space=1&num=1553

Code : https://pastebin.com/kihRmD4Z

![image](https://github.com/user-attachments/assets/485415dd-9439-4733-a216-dc0f6bad6661)

OmAmar106 commented:
> @OmAmar106 Nice work! One suggestion — would it be possible to refactor the code to make it more generic? Specifically, can you modify it to take a function as a parameter (e.g., func), which can then be used to perform different types of queries like sum, max, etc.? This would make the implementation more reusable across different use cases.

Yeah Makes Sense, Will do.

OmAmar106 commented:
@MananRathod I made the Functions as a parameter, along with I have also added the initial value for the query function as a parameter for the HLD. 

![image](https://github.com/user-attachments/assets/16622bd5-73eb-41f8-9c35-a2aaa5891db3)

This is the AC submission to the same problem with the new HLD code, the code used: https://pastebin.com/V59iF9y5
Lemme know if anything else needs changing.

OmAmar106 commented:
> @OmAmar106 I tried to solve the problem - https://cses.fi/problemset/task/1138 . The code was giving wrong answer for the sample test case. Can you Please check. My Output - [11, 13]. Actual Output - [11, 8] Here is the code used - https://pastebin.com/FFk23gHS

The problem is that the segment tree i am using when updates the value, it adds the value rather than updating, to update the value completely, change line 140 in your code from

self.tree[pos] += value

to 

self.tree[pos] = value

this way it assigns the value, but you will still get TLE on 3 of the 11 testcases, due to the tight time constraints of CSES.

Code : https://cses.fi/paste/6a6c8ca588742916c8de1d/

MananRathod commented:
@OmAmar106 Thanks for pointing that out. Changing self.tree[pos] += value to self.tree[pos] = value makes the update operation behave correctly. 
I also tested it using my own segment tree template, and it seems to work.

This kind of update (where we assign the value instead of adding) is actually one of the most common cases I’ve come across, so it might make sense to include it in the base template. If this approach looks good to you, you could update your PR to reflect the change.

@cheran-senthil @Mukundan314 Could you please take a look when you get a chance? If everything looks good, I think this can be a helpful code to merge.

OmAmar106 commented:
@MananRathod True, I changed it to assignment.

OmAmar106 commented:
> I do like the idea of adding heavy light to pyrival. But there are some things that I think should be considered before adding it.
> 
> 1. Is it possible to simplify the code (shorter code/fewer variables)? To me it feels like it should possible, especially with _dfs and _decompose. One idea for shorter query code is to implement LCA. With LCA you should be able to simplify the the query code, and this would give the user the added benifit of a nice and useful LCA function.
> 2. I think it is important to make sure this is running fast in order for it to be useful in practice. While it is true that cses has tight time limits, it would be really good to get it to pass even tight time limits. Here is a good benchmark codeforces problem that can be solved with heavy light https://codeforces.com/contest/343/problem/D . You might need to use the trick from https://codeforces.com/blog/entry/53170 to solve it. Here is my old pypy2 HLD solution (<2s) to that problem https://codeforces.com/contest/343/submission/129004332 .
> 3. Ideally it should be possible to freely swap out which segment tree is being used. For example, suppose you have a segment tree that allows for updates on intervals. It is not super important, but it would be nice if this was easy to do.

@bjorn-martinsson 
Okay will do this, for the third point, maybe i will just change the function to take in variable number of arguments and depending upon whether their are 2(point upd) or 3(range upd) arguments, I will call different functions.

Might take a bit time though, but surely will do.

Also Big Fan 🫡.

OmAmar106 commented:
@bjorn-martinsson 

I added the functionality of Lazily Propogated Segment Tree, so now it will be possible to swap segment trees. I also added two new function, first being update path, it needs a Lazy Segment Tree to update/assign on the path from vertex u to v (passed as parameter). Second being add_to_subtree, it adds/assigns a particular value to the subtree of a vertex. But adding these two functions made the code a bit more longer, but i tried removing as many variable as i could.

As for the problem https://codeforces.com/contest/343/problem/D , I was able to solve it but was not able do it using your rmethod, i ended up using Lazy Segment Tree https://codeforces.com/contest/343/submission/323184524 , I did get AC, but it took 3390ms, since Lazy Segment Tree is quite slow. Lemme know if this works, I will try to improve it more, but might not be able to, seems a bit hard for my level now ;-;.

bjorn-martinsson commented:
I've been thinking lately about the HLD stuff. I think I've found a very nice way of doing it, albeit a bit slow since it uses generators (yield). 

Here are the features of my implementation:
1. It constructs the HL-decomposition iteratively instead of using a stack. The reason I did this is more of an experiment. Probably using a stack like you did is nicer. I haven't benchmarked it.
2. My HLD has no knowledge of what segment tree is being used. This makes it increadibly easy to combine with any segment tree. Instead my HLD has functions like `query_path`, `query_subtree`, `query_node` that return what intervals to give to a segment tree.
3. I've implemented an LCA function. It is nice feature to have.

Here are two submissions to the water problem of my HLD (that use very different segment trees implementations), https://codeforces.com/contest/343/submission/327533352 and  https://codeforces.com/contest/343/submission/327533749 . Note how the solution part of the code is almost identical. I've not have to modify the HLD part of the code at all.

bjorn-martinsson commented:
An update. I tried out my HLD implementation to solve Path Queries 1 and Path Queries 2 on cses (https://cses.fi/problemset/task/1138/ resp. https://cses.fi/problemset/task/2134/).

* On Path Queries 1 it gets AC: https://cses.fi/paste/ad953f8ff3f3e39acf5c37/ (runs in 0.97s / 1.0s).

* On Path Queries 2 it gets AC on all but 3 test cases: https://cses.fi/paste/b1eb4d5c1e058f45cf5c0d/ (note: there are no AC Python submissions for this problem, especially tc9 is almost impossible to AC in Python.)

I'm starting to really like this HLD implementation. It is easy to use, and it seems to be sufficiently fast.

OmAmar106 commented:
> An update. I tried out my HLD implementation to solve Path Queries 1 and Path Queries 2 on cses (https://cses.fi/problemset/task/1138/ resp. https://cses.fi/problemset/task/2134/).
> 
> * On Path Queries 1 it gets AC: https://cses.fi/paste/ad953f8ff3f3e39acf5c37/ (runs in 0.97s / 1.0s).
> * On Path Queries 2 it gets TLE on 3 out of 11 test cases: https://cses.fi/problemset/result/13589633/ (note: there are no AC Python submissions for this problem, especially tc9 is almost impossible to AC in Python.)
> 
> I'm starting to really like this HLD implementation. It is easy to use, and it seems to be sufficiently fast.

Yeah, your seems a lot more neater and faster, I guess I myself will switch over to your implementation, since mine is quite slow, gives TLE on 6 of the Test Cases in Path Queries II.

bjorn-martinsson commented:
> > An update. I tried out my HLD implementation to solve Path Queries 1 and Path Queries 2 on cses (https://cses.fi/problemset/task/1138/ resp. https://cses.fi/problemset/task/2134/).
> > 
> > * On Path Queries 1 it gets AC: https://cses.fi/paste/ad953f8ff3f3e39acf5c37/ (runs in 0.97s / 1.0s).
> > * On Path Queries 2 it gets TLE on 3 out of 11 test cases: https://cses.fi/problemset/result/13589633/ (note: there are no AC Python submissions for this problem, especially tc9 is almost impossible to AC in Python.)
> > 
> > I'm starting to really like this HLD implementation. It is easy to use, and it seems to be sufficiently fast.
> 
> Yeah, your seems a lot more neater and faster, I guess I myself will switch over to your implementation, since mine is quite slow, gives TLE on 6 of the Test Cases in Path Queries II.

Something to watch out for is that on Path Queries II is that I've optimized my segment tree quite a bit. So if you want to compare our HLD implementations, you probably should use the same segment tree as me.

OmAmar106 commented:
@bjorn-martinsson  
Could you share the code for Path Queries II in cses.fi/paste format the way you did for Path Queries I, since https://cses.fi/problemset/result/13589633/ this only works for the one who submitted the code

bjorn-martinsson commented:
> @bjorn-martinsson Could you share the code for Path Queries II in cses.fi/paste format the way you did for Path Queries I, since https://cses.fi/problemset/result/13589633/ this only works for the one who submitted the code

Oh I posted the wrong link by mistake. Here is the correct one for Path Queries II https://cses.fi/paste/201b2376ee4b3f61cf5c81/ .
